src/LICM.cpp

```diff
@@ -346,10 +346,24 @@ class GroupLoopInvariants : public IRMutator2 {
     }
 
     Expr visit(const Sub *op) override {
+        if (op->type.is_float()) {
+            // Don't reassociate float exprs; it doesn't play well with -ffast-math.
+            // TODO: It might be safe to do this if 'strict_float' is enabled;
```

Comment:
Wouldn't the logic be that we would do this if `strict_float` exists in the language but it is *not* present? My understanding is that strict_float -> don't reassociate.

Reply:
Yes, I garbled the comment after a long day of debugging. (And TBH, I'm not sure it's a good idea even then, since it could produce -0.0 which couldn't be dealt with reliably if strict_float is off.) I'll reword the comment.

Reply:
Without StrictFP, the failing test is just wrong. We do floating point reassociation elsewhere, and so does llvm, and the test is unstable to that. We're only backing it out because FP reassociation is not necessary here and it causes a minor headache even though we're technically in the right.

Reply:
Yes, but without StrictFP, I'm not sure if the failing test can easily be made right, since checking for signed-zero in that case is not reliable. (Though I suppose the right answer would be to do an epsilon-based comparison...)

(Background for readers: the test was reassociating a floating-point summation and ended up producing -0.0 instead of 0.0; this caused a select() that checked for "product < 0" to flip depending on optimization vagaries.)